---
title: "Introduction and Examples of Use"
author: "Peter Dolan"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(dtou)
library(kableExtra)
```

## Purpose

The purpose of this package is to aid in the rapid prototyping of analytic pipelines by introducing an algorithm developed to calculate the **distance to uniqueness** (see below) metric of genomic sequences into R.  The package implements the `dtou` function-- uses of which are shown below, but perhaps even more importantly, the algorithm implemented in the C++ source code that forms the backbone of the procedure can be modified for novel analytic purposes (see [companion vignette for modifying source code](AlgorithmModifications.html) for an example)

## Distance to Uniqueness

The distance to uniqueness for a base-pair in a collection of genomic sequences is defined for each base as the shortest unique subsequence starting at that base.  The reverse complement can be taken into consideration when deciding the meaning of a unique subsequence.  It's easiest to see with a simple example that finds the metric for the base-pairs in the string `"AAAAACCCGACTGGGCTCA"`:

```{r}
str="AAAAACCCGACTGGGCTCA"
(results<-dtou(str,rc=TRUE)[[1]])
```

(The default value for `rc` is `TRUE`, but its nice to be explicit)

It will be easier to explain with a table:

```{r,echo=FALSE}
tmp=strsplit(str,'')[[1]]
names(results)=tmp
kable(t(results))
```

The 5 under the initial `A` means that the 4-mer `AAAA` is repeated, but the 5-mer `AAAAA` is not.  This is because overlaps are allowed.

The 4 under the first `C` means that the 4-mer `CCCG` is unique but the 3-mer `CCC` is not.  Although the 3-mer `CCC` is not repeated, its reverse complement `GGG` does appear in the string.  If we run `dtou()` with `rc=FALSE`:

```{r}
(results<-dtou(str,rc=FALSE)[[1]])
```

```{r,echo=FALSE}
tmp=strsplit(str,'')[[1]]
names(results)=tmp
kable(t(results))
```

Then `CCC` is unique and the metric has changed to represent this fact.

## Usage

The workhorse functions are all written in C++.  They *can* be invoked directly, but the function `dtou()` is designed to find the appropriate function based upon its arguments. The following arguments determine the behavior of the function

Argument  | Influence
----------|----------
`rc`      | Boolean that determines if the reverse comlement is taken into consideration
`depth`   | By default the recursion is not depth-limited.  Setting this value to a positive number determines the max depth of the recursion
`optimizeForSpeed` | Determines whether the algorithm switches to a faster technique when the stack size is 2

## Ecoli example

I downloaded and ecoli genome (strain K12, substrain MG 1655 ) with assembly name ASM584v2 using [NCBI Reference Sequence NC_000913.3](https://www.ncbi.nlm.nih.gov/nuccore/NC_000913.3?report=fasta).

When this vignette was compiled the link above leads to a webpage from which the fasta file may be downloaded.  I named my copy `ecoli_str_k_12_substr_mg1655.ASM584v2.fa` and put it in a subdirectory called `data`.  I also used `Biostrings::readDNAStringSet()` to load the data and then convert it to a character vector.  This can be done without the use of `Biostrings`, but Biostrings makes things much easier.  See the [Bioconductor page for Biostrings](https://bioconductor.org/packages/release/bioc/html/Biostrings.html) for information on how to install `BioStrings`:

```{r,echo=FALSE,output=FALSE,message=FALSE}
library(Biostrings)
ecoli=readDNAStringSet("../data/ecoli_str_k_12_substr_mg1655.ASM584v2.fa")
```

```{r,eval=FALSE}
library(Biostrings)
ecoli=readDNAStringSet("data/ecoli_str_k_12_substr_mg1655.ASM584v2.fa")
```


```{r}
str<-as.character(ecoli)
system.time({ecoli.dtou<-dtou(str)[[1]]})
```

Let's start with a graph of the metric for the genome:

```{r,fig.width=7}
plot(ecoli.dtou,
		 pch=".",
		 main="Distance to Uniqueness for base-pairs in Ecoli sequence",
		 xlab="index",ylab="dtou"
)
```

From which we can see that there are some regions of high repetition spread-out somewhat evenly along the genome.  The two highest peaks are likely the longest repeated region.  The second highest peak is likely due to substantial overlap with the first peaks, but the scale of the index along the bottom is not consistent with that along the top-- the 3000 on the y-axis only covers 1/3 of a percent of the distance from 1e+06 to 2e+06.  

We can find the maximum peak and have a closer look:

```{r,fig.width=7}
window.length=7000
i=which.max(ecoli.dtou)
lower=i-window.length
upper=i+window.length
zoom=lower:upper
head(zoom)
plot(zoom,
		 ecoli.dtou[zoom],
		 pch=".",
		 main="Distance to Uniqueness for base-pairs in Ecoli sequence (zoomed)",
		 xlab="index",ylab="dtou"
)
```

The majority of the dtou's for this sequence are low so it's more informative to look at the deciles than a histogram:

```{r}
quantile(ecoli.dtou,1:9*0.1)
```

We can see that only 2.8 percent of the genome is repetitive at a $20$-mer level or higher:

```{r}
sum(ecoli.dtou>20)
sum(ecoli.dtou>20)/length(ecoli.dtou)
```

This histogram of this shows an interesting distribution:

```{r,fig.width=7}
hist(log10(ecoli.dtou[ecoli.dtou>20]),main="E.coli Upper Tail Distance to Uniqueness",xlab="log10 dtou",sub="dtou > 20")
```

We can get a closely related substrain as well:

```{r,echo=FALSE,output=FALSE,message=FALSE}
ecoli2=readDNAStringSet("../data/GCF_000005845.2_ASM584v2_genomic.fna")
```
```{r,eval=FALSE}
ecoli2=readDNAStringSet("data/GCF_000005845.2_ASM584v2_genomic.fna")
str2<-as.character(ecoli2)
system.time({ecoli2.dtou<-dtou(str2)[[1]]})
```

And 

# Homo Sapiens Chromosome I

Here we will download *Homo Sapiens* Chromomome I and calculate the distance to uniqueness for the entire chromosome.  Appropriate modifications will need to be made if this example is being run using the windows command prompt instead of bash:

```{bash,eval=FALSE}
cd _bigTest
wget ftp://ftp.ncbi.nih.gov/genomes/Homo_sapiens/CHR_01/hs_ref_GRCh38.p12_chr1.fa.gz
gunzip ./hs_ref_GRCh38.p12_chr1.fa.gz
```

```{r,eval=FALSE}
library(Biostrings)
chr1<-readDNAStringSet("../_bigTest/hs_ref_GRCh38.p12_chr1.fa",format="fasta")
```

The alternate scaffolds are repeats.. so let's remove them:

```{r,eval=FALSE}
chr1<-chr1[!grepl("alternate",names(chr1))]
str=as.character(chr1)
```

Now let's do some timing:

```{r,eval=FALSE}
system.time({results<-c_limiteddtouC(str,10000)})
```


