---
title: "Vignette Title"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))

#Vignette proper

## Ecoli example

Let's look at a quick example using ecoli.  

Start by downloading the file.  It can be 

```{bash,eval=FALSE}

```

```{r,eval=FALSE}
ecoli=readDNAStringSet("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna_sm.chromosome.Chromosome.fa")
length(ecoli)
str=as.character(ecoli)
Rcpp::sourceCpp("src/dtou3.cpp")
system.time({ecoli.results<-c_dtouC(str)})
quantile(ecoli.results[[1]],1:99*0.01)
sum(ecoli.results[[1]]>20)
sum(ecoli.results[[1]]>20)/length(ecoli.results[[1]])
```




Here we will download *Homo Sapiens* Chromomome I and calculate the distance to uniqueness for the entire chromosome.  Appropriate modifications will need to be made if this example is being run using the windows command prompt instead of bash:

```{bash,eval=FALSE}
cd _bigTest
wget ftp://ftp.ncbi.nih.gov/genomes/Homo_sapiens/CHR_01/hs_ref_GRCh38.p12_chr1.fa.gz
gunzip ./hs_ref_GRCh38.p12_chr1.fa.gz
```

The next part doesn't *require* `Biostrings`, but it's much easier 
```{r,eval=FALSE}
library(Biostrings)
chr1<-readDNAStringSet("../_bigTest/hs_ref_GRCh38.p12_chr1.fa",format="fasta")
```

The alternate scaffolds are repeats.. so let's remove them:

```{r,eval=FALSE}
chr1<-chr1[!grepl("alternate",names(chr1))]
str=as.character(chr1)
```

Now let's do some timing:

```{r,eval=FALSE}
system.time({results<-c_limiteddtouC(str,10000)})
```


