---
title: "Vignette"
author: "Peter Dolan"
date: "9/24/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Here we will download *Homo Sapiens* Chromomome I and calculate the distance to uniqueness for the entire chromosome.  Appropriate modifications will need to be made if this example is being run using the windows command prompt instead of bash:

```{bash,eval=FALSE}
cd _bigTest
wget ftp://ftp.ncbi.nih.gov/genomes/Homo_sapiens/CHR_01/hs_ref_GRCh38.p12_chr1.fa.gz
gunzip ./hs_ref_GRCh38.p12_chr1.fa.gz
```

The next part doesn't *require* `Biostrings`, but it's much easier 
```{r}
library(Biostrings)
chr1<-readDNAStringSet("_bigTest/hs_ref_GRCh38.p12_chr1.fa",format="fasta")
```

The alternate scaffolds are repeats.. so let's remove them:

```{r}
chr1<-chr1[!grepl("alternate",names(chr1))]
str=as.character(chr1)
```

Now let's do the loading

```{r}
require(Rcpp)
Rcpp::sourceCpp("src/dtou3withlimits.cpp")
```

```{r}
system.time({results<-c_limiteddtouC(str,10000)})
```

Let's look at a quick example using ecoli

```{r,eval=FALSE}
ecoli=readDNAStringSet("Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.dna_sm.chromosome.Chromosome.fa")
length(ecoli)
str=as.character(ecoli)
Rcpp::sourceCpp("src/dtou3.cpp")
system.time({ecoli.results<-c_dtouC(str)})
quantile(ecoli.results[[1]],1:99*0.01)
sum(ecoli.results[[1]]>20)
sum(ecoli.results[[1]]>20)/length(ecoli.results[[1]])
```

